---
title: "Neotropical Metabarcoding Metadata"
author: "Frédéric Mahé"
date: '`r format(Sys.time(), "%d %B %Y")`'

output:
  rmarkdown::html_document:
    theme: lumen
    toc: yes
    toc_float: TRUE
    keep_md: yes
    # code_folding: hide
---

```{r setup, include=FALSE}
rm(list = ls()) # remove all objects before starting
knitr::opts_chunk$set(echo = TRUE)
```


#### load required packages

```{r packages, message=FALSE}
library(here)
library(tidyverse)
library(lubridate)
```

# preliminary steps

### variables and functions

```{r}
eukbank_metadata_file <- "eukbank_18SV4_asv.metadata"
coordinates_file <- "sample_coordinates.tsv"
project_name_in_eukbank <- "Neotropical_lowland_forests"
list_of_read_files <- "list_of_read_files.tsv"
remove_days <- lubridate::stamp("2019-02", orders = "ym")
roche_454_samples <- c("B010", "B020", "B030", "B040", "B050", "B060", "B070", "B080", "B090", "B100", "L010", "L020", "L030", "L040", "L050", "L060", "L070", "L080", "L090", "L100")
output <- "neotrop_samples.tsv"
```

```{r}
load_list_of_read_files <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(sample = str_remove(string = Submitted_files,
                                    pattern = "_lib.*"),
               sample = str_remove(string = sample,
                                    pattern = "NG-[0-9]+_")) %>%
        filter(str_detect(string = Submitted_files,
                          pattern = "_1_2\\.",
                          negate = TRUE))
}


load_per_sample_coordinates <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(northing = as.character(northing),
               sample_description = str_glue(
                   "Sampling position of {sample}",
                   "in UTM coordinates is zone: {zone},",
                   "northing: {northing}, easting: {easting}.",
                   .sep = " ")
               )
}


load_raw_metadata <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        filter(project == project_name_in_eukbank) %>%
        select(-c(project, notus, nreads, size_fraction_lower_threshold,
                  size_fraction_upper_threshold))
}


. %>%
    mutate(sample = case_when(
               sample == "T185_186" ~ "T185_T186",
               sample == "L137_L13" ~ "L137_L138",
               TRUE                 ~ sample),
           collection_date = case_when(
               sample == "T185_T186" ~ "2013-10-01",
               sample == "T199_T200" ~ "2013-10-01",
               sample == "L137_L138" ~ "2013-06-01",
               TRUE                  ~ collection_date),
           longitude = case_when(
               sample == "T185_T186" ~ -76.1690818455638,
               sample == "T199_T200" ~ -76.1583931324974,
               sample == "L137_L138" ~ -84.0081054902495,
               TRUE                  ~ longitude),
           latitude = case_when(
               sample == "T185_T186" ~ -0.622442092820539,
               sample == "T199_T200" ~ -0.61982581193797,
               sample == "L137_L138" ~ 10.4242830027919,
               TRUE                  ~ latitude)
           ) -> fix_missing_data
```


### get a list of available read files (sff and fastq)

```{r}
load_list_of_read_files(list_of_read_files) -> read_file_names
```


Some samples were sequenced twice (Roche 454 and Illumina
MiSeq). After extracting sample names from read file names, we expect
to find 154 unique sample names.

```{r}
read_file_names %>%
    distinct(sample) %>%
    pull() %>%
    length() -> n_unique_samples
```

There are `r prettyNum(n_unique_samples, scientific=FALSE,
big.mark=",")` unique sample names.


### get per-sample coordinates

Using the Universal Transverse Mercator (UTM) coordinate system:

```{r}
load_per_sample_coordinates(coordinates_file) -> sample_cordinates
```

# load metadata

```{r}
load_raw_metadata(eukbank_metadata_file) %>%
    fix_missing_data -> metadata
```

## check

Are sample lists equal?

```{r}
setequal(x = metadata %>%
             distinct(sample),
         y = read_file_names %>%
             distinct(sample))
```

Yes, the metadata table and the list of available read files contain
the same lists of sample names.


## create the 'sample description' field

```{r}
## generate sample description for double samples
. %>%
    select(sample) %>%
    filter(str_detect(string = sample, pattern = "_")) %>%
    mutate(sample_pair = sample) %>%
    separate_rows(sample, sep = "_") %>%
    left_join(sample_cordinates %>%
              select(sample, sample_description),
              by = "sample") %>%
    select(-sample) %>%
    group_by(sample_pair) %>%
    mutate(position = row_number(sample_pair)) %>%
    ungroup() %>%
    pivot_wider(names_from = position, values_from = sample_description) %>%
    unite(2, 3, col = "sample_description", sep = " ") %>%
    mutate(sample_description = as.character(sample_description)) %>%
    rename(sample = sample_pair) -> sample_descriptions_for_double_samples

## bind with sample descriptions for single samples
bind_rows(metadata %>%
          sample_descriptions_for_double_samples,
          sample_cordinates %>%
          select(sample, sample_description)
          ) -> sample_descriptions
```


## create and order columns as required in the ENA UniEuk_EukBank Checklist

```{r}
metadata %>%
    left_join(x = ., y = sample_descriptions, by = "sample") %>%
    mutate(tax_id = 100272,
           scientific_name = "uncultured eukaryote",
           sample_alias = sample,
           sample_title = case_when(
               str_starts(string = sample, pattern = "L") ~ "soil sample from La Selva Biological Station, Costa Rica",
               str_starts(string = sample, pattern = "B") ~ "soil sample from Barro Colorado Island, Panama",
               str_starts(string = sample, pattern = "T") ~ "soil sample from Tiputini Biodiversity Station, Ecuador",
               TRUE ~ "unknown"
           ),
           sample_description = as.character(sample_description),
           `target gene` = "18S",
           `target subfragment` = "V4",
           `pcr primers` = primers,
           isolation_source	= "forest top-soil",
           collected_by = "Micah Dunthorn",
           collection_date = lubridate::parse_date_time(collection_date, "ymd"),
           collection_date = remove_days(collection_date),
           `collection date` = collection_date,
           `geographic location (country and/or sea)` = case_when(
               str_starts(string = sample, pattern = "L") ~ "Costa Rica",
               str_starts(string = sample, pattern = "B") ~ "Panama",
               str_starts(string = sample, pattern = "T") ~ "Ecuador",
               TRUE ~ "unknown"
           ),
           `geographic location (latitude)` = latitude,
           `geographic location (longitude)` = longitude,
           `geographic location (region and locality)` = case_when(
               str_starts(string = sample, pattern = "L") ~ "La Selva Biological Station",
               str_starts(string = sample, pattern = "B") ~ "Barro Colorado Island",
               str_starts(string = sample, pattern = "T") ~ "Tiputini Biodiversity Station",
               TRUE ~ "unknown"
           ),
           depth = 0,
           `broad-scale environmental context` = biome,
           `local environmental context` = feature,
           `environmental medium` = material,
           `sample collection device or method` = "spoon",
           environmental_sample = "yes",
           `Further Details` = "http://unieuk.org/project/",
           .before = 1) %>%
    relocate(sample_description, .after = sample_title) %>%
    relocate(depth, .after = `geographic location (region and locality)`) %>%
    select(tax_id:`Further Details`) %>%
    arrange(sample_alias) %>%
    write_tsv(here::here("results", output))
```

**TODO**:

- **DONE** fix missing data (coordinates and sampling dates),
- extract `column_generator` from the above block (i.e. the whole `mutate`),
- make a second table creator,
- update wrong values on-the-fly (`tax_id`, `scientific_name`, `Further Details`, `target subfragment`, and `pcr primers`)
- bind the two tables


***

```{r}
sessionInfo()
rm(list = ls())
```
