---
title: "Neotropical Metabarcoding Metadata"
author: "Frédéric Mahé"
date: '`r format(Sys.time(), "%d %B %Y")`'

output:
  rmarkdown::html_document:
    theme: lumen
    toc: yes
    toc_float: TRUE
    keep_md: yes
    # code_folding: hide
---

```{r setup, include=FALSE}
rm(list = ls()) # remove all objects before starting
knitr::opts_chunk$set(echo = TRUE)
```


#### load required packages

```{r packages, message=FALSE}
library(here)
library(tidyverse)
library(lubridate)
```


# preliminary steps

### variables and functions

```{r}
eukbank_metadata_file <- "eukbank_18SV4_asv.metadata"
coordinates_file <- "sample_coordinates.tsv"
project_name_in_eukbank <- "Neotropical_lowland_forests"
list_of_read_files <- "list_of_read_files.tsv"
remove_days <- lubridate::stamp("2019-02", orders = "%Y-%Om", quiet = TRUE)
roche_454_samples <- c("B010", "B020", "B030", "B040", "B050", "B060", "B070",
                       "B080", "B090", "B100", "L010", "L020", "L030", "L040",
                       "L050", "L060", "L070", "L080", "L090", "L100")
La_Selva <- "La Selva Biological Station"
Barro <- "Barro Colorado Island"
Tiputini <- "Tiputini Biodiversity Station"
La_Selva_country <- "Costa Rica"
Barro_country <- "Panama"
Tiputini_country <- "Ecuador"
title_prefix <- "soil sample from "
La_Selva_title <- str_c(title_prefix, La_Selva, ", ", La_Selva_country)
Barro_title <- str_c(title_prefix, Barro, ", ", Barro_country)
Tiputini_title <- str_c(title_prefix, Tiputini, ", ", Tiputini_country)
output <- "neotrop_samples.tsv"
output_ncbi <- "neotrop_samples_ncbi.tsv"
```

```{r}
load_list_of_read_files <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(sample = str_remove(string = Submitted_files,
                                    pattern = "_lib.*"),
               sample = str_remove(string = sample,
                                    pattern = "NG-[0-9]+_")) %>%
        filter(str_detect(string = Submitted_files,
                          pattern = "_1_2\\.",
                          negate = TRUE))
}


load_per_sample_coordinates <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(northing = as.character(northing),
               sample_description = str_glue(
                   "Sampling position of {sample}",
                   "in UTM coordinates is zone: {zone},",
                   "northing: {northing}, easting: {easting}.",
                   .sep = " ")
               )
}


load_raw_metadata <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        filter(project == project_name_in_eukbank) %>%
        select(-c(project, notus, nreads, size_fraction_lower_threshold,
                  size_fraction_upper_threshold))
}


. %>%
    mutate(sample = case_when(
               sample == "T185_186" ~ "T185_T186",
               sample == "L137_L13" ~ "L137_L138",
               TRUE                 ~ sample),
           collection_date = case_when(
               sample == "T185_T186" ~ "2013-10-01",
               sample == "T199_T200" ~ "2013-10-01",
               sample == "L137_L138" ~ "2013-06-01",
               TRUE                  ~ collection_date),
           longitude = case_when(
               sample == "T185_T186" ~ -76.1690818455638,
               sample == "T199_T200" ~ -76.1583931324974,
               sample == "L137_L138" ~ -84.0081054902495,
               TRUE                  ~ longitude),
           latitude = case_when(
               sample == "T185_T186" ~ -0.622442092820539,
               sample == "T199_T200" ~ -0.61982581193797,
               sample == "L137_L138" ~ 10.4242830027919,
               TRUE                  ~ latitude)
           ) -> fix_missing_data
```


### get a list of available read files (sff and fastq)

```{r}
load_list_of_read_files(list_of_read_files) -> read_file_names
```


Some samples were sequenced twice (Roche 454 and Illumina
MiSeq). After extracting sample names from read file names, we expect
to find 154 unique sample names.

```{r}
read_file_names %>%
    distinct(sample) %>%
    pull() %>%
    length() -> n_unique_samples
```

There are `r prettyNum(n_unique_samples, scientific=FALSE,
big.mark=",")` unique sample names.


### get per-sample coordinates

Using the Universal Transverse Mercator (UTM) coordinate system:

```{r}
load_per_sample_coordinates(coordinates_file) -> sample_cordinates
```


# load metadata

```{r}
load_raw_metadata(eukbank_metadata_file) %>%
    fix_missing_data -> metadata
```


## check

Are sample lists equal?

```{r}
setequal(x = metadata %>%
             distinct(sample),
         y = read_file_names %>%
             distinct(sample))
```

after fixing two sample names, the metadata table and the list of
available read files now differ.


## create the 'sample description' field

generate sample description for double samples:

```{r}
. %>%
    select(sample) %>%
    filter(str_detect(string = sample, pattern = "_")) %>%
    mutate(sample_pair = sample) %>%
    separate_rows(sample, sep = "_") %>%
    left_join(sample_cordinates %>%
              select(sample, sample_description),
              by = "sample") %>%
    select(-sample) %>%
    group_by(sample_pair) %>%
    mutate(position = row_number(sample_pair)) %>%
    ungroup() %>%
    pivot_wider(names_from = position, values_from = sample_description) %>%
    unite(2, 3, col = "sample_description", sep = " ") %>%
    mutate(sample_description = as.character(sample_description)) %>%
    rename(sample = sample_pair) -> sample_descriptions_for_double_samples
```

bind with sample descriptions for single samples:

```{r}
bind_rows(metadata %>%
          sample_descriptions_for_double_samples,
          sample_cordinates %>%
          select(sample, sample_description)
          ) -> sample_descriptions
```


# ENA UniEuk_EukBank Checklist

Create and order columns as specified in the [checklist
ERC000040](https://www.ebi.ac.uk/ena/browser/view/ERC000040):

```{r}
. %>%
    mutate(tax_id = 342097,
           scientific_name = "uncultured soil eukaryote",
           sample_alias = sample,
           sample_title = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva_title,
               str_starts(string = sample, pattern = "B") ~ Barro_title,
               str_starts(string = sample, pattern = "T") ~ Tiputini_title,
               TRUE ~ "unknown"
           ),
           sample_description = as.character(sample_description),
           `target gene` = "18S SSU rRNA",
           `target subfragment` = "V4",
           `pcr primers` = primers,
           isolation_source	= "forest top-soil",
           collected_by = "Micah Dunthorn",
           collection_date = lubridate::parse_date_time(collection_date, "ymd"),
           collection_date = remove_days(collection_date),
           `collection date` = collection_date,
           `geographic location (country and/or sea)` = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva_country,
               str_starts(string = sample, pattern = "B") ~ Barro_country,
               str_starts(string = sample, pattern = "T") ~ Tiputini_country,
               TRUE ~ "unknown"
           ),
           `geographic location (latitude)` = latitude,
           `geographic location (longitude)` = longitude,
           `geographic location (region and locality)` = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva,
               str_starts(string = sample, pattern = "B") ~ Barro,
               str_starts(string = sample, pattern = "T") ~ Tiputini,
               TRUE ~ "unknown"
           ),
           depth = 0,
           `broad-scale environmental context` = biome,
           `local environmental context` = feature,
           `environmental medium` = material,
           `sample collection device or method` = "spoon",
           environmental_sample = "Yes",
           `Further Details` = "http://unieuk.org/project/",
           .before = 1) -> create_eukbank_columns
```

Which level of precision should we choose?

- `tax_id` = 100272, `scientific_name` = "uncultured eukaryote"
- `tax id` = 342097, `scientific_name` = "uncultured soil eukaryote"

Sadly, "uncultured soil eukaryote" is not a sub-node of "uncultured
eukaryote". Nonetheless, I think its is best to use the most precise
expression.

([Environment Ontology](https://ontobee.org/ontology/ENVO))


## Illumina and Roche 454 samples (universal primers)

The primers are:

- V4F = CCAGCASCYGCGGTAATTCC
- V4R = ACTTTCGTTCTTGATYRA

Also known as:

- TAReuk454FWD1 (Forward): 5‘-CCAGCA(G/C)C(C/T)GCGGTAATTCC-3‘
- TAReukRev3 (Reverse): 5‘-ACTTTCGTTCTTGAT(C/T)(A/-G)A-3‘

Standard master mix (in μl):

- 19,75 H20
- 2,5 10x Buffer
- 0,5 dNTP
- 0,5 V4F
- 0,5 V4R
- 0,25 polymerase
- 1 template

Cycling conditions:

(with Hot Start Taq from Qiagen)

1) 95°C 5 min
2) 94°C 30 sec
3) 47°C 45 sec
4) 72°C 60 sec; 2-4 = 29x
5) 72°C 5 min


```{r}
metadata %>%
    left_join(x = ., y = sample_descriptions, by = "sample") %>%
    create_eukbank_columns %>%
    relocate(sample_description, .after = sample_title) %>%
    relocate(depth, .after = `geographic location (region and locality)`) %>%
    select(tax_id:`Further Details`) %>%
    arrange(sample_alias) -> universal_primer_samples
```


## Roche 454 samples (ciliate-specific primers, followed by universal primers)

20 samples, using a two-step amplification:

1) ciliate specific primers,
2) universal V4 primers

The primers are:

- Cil_F: 5’ - TGG TAG TGT ATT GGA CWA CCA -3’

Cil R (an equimolar mix of three different primers):

- 5’ - TCT GAT CGT CTT TGA TCC CTT A – 3’
- 5’ - TCT RAT CGT CTT TGA TCC CCT A – 3’
- 5’ - TCT GAT TGT CTT TGA TCC CCT A – 3’

Cycling conditions:

(with Hot Start Taq from Qiagen)

1) 95°C 5 min
2) 94°C 30 sec
3) 56°C 30 sec
4) 72°C 60 sec; 2-4 = 25x
5) 72°C 10 min

followed by an amplification with universal 18S SSU rRNA V4 primers
(see above).

```{r}
message <- "two-step amplification: 1) ciliate specific primers, 2) universal V4 primers (may contain taxonomic groups other than ciliates)"

universal_primer_samples %>%
    filter(sample_alias %in% roche_454_samples) %>%
    mutate(tax_id = 187299,
           scientific_name = "uncultured ciliate",
           `Further Details` = message) -> ciliate_specific_samples
```


## save results

```{r}
bind_rows(universal_primer_samples, ciliate_specific_samples) %>%
    write_tsv(here::here("results", output))
```

Should the second set be submitted separately? without the ERC000040
ENA UniEuk_EukBank Checklist? EukBank 1.0 does not use the Ciliate
samples, but it might in the future, so to make things simpler, I
should submit only one table.


# NCBI's template (biosamples)

Submission template for batch deposit of 'MIMARKS: survey, soil;
version 5.0' samples to the NCBI BioSample database
(https://www.ncbi.nlm.nih.gov/biosample/).

Use for any type of marker gene sequences, eg, 16S, 18S, 23S, 28S rRNA
or COI obtained directly from the environment, without culturing or
identification of the organisms. Organism must be a metagenome, where
lineage starts with unclassified sequences and scientific name ends
with 'metagenome'.

## list of fields

Fields with an asterisk (*) are mandatory. Your submission will fail
if any mandatory fields are not completed. If information is
unavailable for any mandatory field, please enter 'not collected',
'not applicable' or 'missing' as appropriate.

All other fields are optional. Leave optional fields empty if no
information is available.

You can add any number of custom fields to fully describe your
BioSamples, simply include them in the table.

https://www.ncbi.nlm.nih.gov/biosample/docs/attributes/

- *sample_name
- sample_title
- bioproject_accession
- *organism
- *collection_date
- *depth
- *elev
- *env_broad_scale
- *env_local_scale
- *env_medium
- *geo_loc_name
- *lat_lon 38.98 N 77.11 W
- isolation_source
- pool_dna_extracts
- samp_collect_device
- samp_mat_process
- samp_size
- description

other potential attributes:

- misc_param
- source_material_id
- collection_method


## prepare table

Exclude ciliate amplifications. Here, the focus is on the
biosamples. Several sequencing runs can be attached to a given
biosample.

```{r}
universal_primer_samples %>%
    mutate(lon_hemisphere  = case_when(
               `geographic location (longitude)` < 0 ~ "W",
               TRUE                                  ~ "E"),
           lat_hemisphere  = case_when(
               `geographic location (latitude)` < 0 ~ "S",
               TRUE                                 ~ "N")) %>%
    mutate(sample_name = sample_alias,
           bioproject_accession = "PRJNA317860",
           organism = "soil metagenome",
           collection_date = `collection date`,
           depth = "0.01 m",
           elev = "not collected",
           env_broad_scale = `broad-scale environmental context`,
           env_local_scale = `local environmental context`,
           env_medium = `environmental medium`,
           geo_loc_name = str_c(`geographic location (country and/or sea)`,
                                `geographic location (region and locality)`,
                                sep = ": "),
           lat_lon = str_c(abs(`geographic location (latitude)`),
                           lat_hemisphere,
                           abs(`geographic location (longitude)`),
                           lon_hemisphere,
                           sep = " "),
           pool_dna_extracts = case_when(
               str_detect(string = sample_name, pattern = "_") ~ "YES;2 extracts",
               TRUE                                            ~ "NO"),
           samp_collect_device = `sample collection device or method`,
           samp_mat_process = "LifeGuard Soil Preservation Solution (8 mL)",
           samp_size = "5 mL",
           env_medium = `environmental medium`,
           description = sample_description,
           .before = 1) %>%
    relocate(collected_by, .after = collection_date) %>%
    relocate(depth, .after = collection_date) %>%
    relocate(isolation_source, .after = lat_lon) %>%
    select(sample_name:description) %>%
    write_tsv(here::here("results", output_ncbi))
```


## submission

To make a submission:
  1. Complete the template table (typically in Excel, or another spreadsheet application)
  2. Save the worksheet as a Text (Tab-delimited) file - use 'File, Save as, Save as type: Text (Tab-delimited)'
  3. Upload the file on the 'Attributes' tab of the BioSample Submission Portal at https://submit.ncbi.nlm.nih.gov/subs/biosample/.
  4. If you have any questions, please contact us at biosamplehelp@ncbi.nlm.nih.gov.

Note: geographical coordinates were automatically truncated after the
6th decimal digit upon submission.

Processing took a few minutes.


# NCBI's template (sequencing runs)

## variables and functions

```{r}
normal_pcr <- str_c(
    c("PCR primers are V4F = CCAGCASCYGCGGTAATTCC and V4R = ",
      "ACTTTCGTTCTTGATYRA, also known as TAReuk454FWD1 (forward) ",
      "5'-CCAGCA(G/C)C(C/T)GCGGTAATTCC-3' and TAReukRev3 (reverse) ",
      "5'-ACTTTCGTTCTTGAT(C/T)(A/-G)A-3'.  PCR mix is (volumes in ",
      "microliters) 19.75 H20, 2.5 10x Buffer, 0.5 dNTP, 0.5 V4F, ",
      "0.5 V4R, 0.25 polymerase, 1 template. Cycling conditions ",
      "(with Hot Start Taq from Qiagen) are 1) 95°C 5 min, 2) 94°C ",
      "30 sec, 3) 47°C 45 sec, 4) 72°C 60 sec, repeat 2-4 = 29x, 5) ",
      "72°C 5 min."),
    collapse = "")

ciliate_pcr <- str_c(
    c("Two-step amplification with 1) ciliate specific primers, and 2) ",
      "universal V4 primers.  The ciliate specific primers are Cil_F 5' ",
      "- TGG TAG TGT ATT GGA CWA CCA -3' and Cil R (an equimolar mix of ",
      "three different primers): 5' - TCT GAT CGT CTT TGA TCC CTT A – 3', ",
      "5' - TCT RAT CGT CTT TGA TCC CCT A – 3' and 5' - TCT GAT TGT CTT ",
      "TGA TCC CCT A – 3'.  Cycling conditions (with Hot Start Taq from ",
      "Qiagen) are 1) 95°C 5 min, 2) 94°C 30 sec, 3) 56°C 30 sec ,4) 72°C ",
      "60 sec, repeat 2-4 = 25x, 5) 72°C 10 min.  Followed by an amplification",
      "with universal 18S SSU rRNA V4 primers (see other sequencing runs in ",
      "the same bioproject)."),
    collapse = "")

project_title <- str_c(
    c("soil eukaryotic 18S SSU rRNA diversity ",
      "in three neotropical broadleaf forests"),
    collapse = "")
```

## load validated table of biosamples

```{r}
here::here("results", "attributes_ncbi.tsv") %>%
    read_tsv(show_col_types = FALSE) %>%
    select(accession, sample_name) %>%
    rename(biosample_accession = accession) -> table_of_biosamples
```


## preload SRA data

see labbook notes on how to load fastq and sff files to the NCBI SRA
FTP server


## build SRA table

```{r}
read_file_names %>%
    mutate(library_number = str_extract(string = Submitted_files,
                                        pattern = "lib[[:digit:]]+(_[[:digit:]]+)?"),
           library_ID = str_c(sample, library_number, sep = "_"),
           title = project_title,
           library_strategy = "AMPLICON",
           library_source = "METAGENOMIC",
           library_selection = "PCR",
           library_layout = case_when(
               str_detect(string = run, pattern = "Roche") ~ "single",
               TRUE                                        ~ "paired"),
           platform = case_when(
               str_detect(string = run, pattern = "Roche") ~ "LS454",
               TRUE                                        ~ "ILLUMINA"),
           instrument_model = case_when(
               str_detect(string = run, pattern = "Roche") ~ "454 GS FLX Titanium",
               TRUE                                        ~ "Illumina MiSeq"),
           design_description = case_when(
               str_detect(string = run, pattern = "Ciliate") ~ ciliate_pcr,
               TRUE                                          ~ normal_pcr),
           filetype = case_when(
               str_detect(string = run, pattern = "Roche") ~ "sff",
               TRUE                                        ~ "fastq"),
           filename = Submitted_files,
           filename2 = case_when(
               filetype == "fastq" ~ str_replace(string = filename,
                                                 pattern = "_1_1",
                                                 replacement = "_1_2"),
               TRUE                ~ ""),
           .before = 1) %>%
    left_join(y = table_of_biosamples, by = c("sample" = "sample_name")) %>%
    relocate(biosample_accession, .before = 1) %>%
    select(-library_number, -run, -Submitted_files, -sample) %>%
    write_tsv(here::here("results", "SRA_submission.tsv"))
```


***

```{r}
sessionInfo()
rm(list = ls())
```
