---
title: "Neotropical Metabarcoding Metadata"
author: "Frédéric Mahé"
date: '`r format(Sys.time(), "%d %B %Y")`'

output:
  rmarkdown::html_document:
    theme: lumen
    toc: yes
    toc_float: TRUE
    keep_md: yes
    # code_folding: hide
---

```{r setup, include=FALSE}
rm(list = ls()) # remove all objects before starting
knitr::opts_chunk$set(echo = TRUE)
```


#### load required packages

```{r packages, message=FALSE}
library(here)
library(tidyverse)
library(lubridate)
```


# preliminary steps

### variables and functions

```{r}
eukbank_metadata_file <- "eukbank_18SV4_asv.metadata"
coordinates_file <- "sample_coordinates.tsv"
project_name_in_eukbank <- "Neotropical_lowland_forests"
list_of_read_files <- "list_of_read_files.tsv"
remove_days <- lubridate::stamp("2019-02", orders = "%Y-%Om", quiet = TRUE)
roche_454_samples <- c("B010", "B020", "B030", "B040", "B050", "B060", "B070",
                       "B080", "B090", "B100", "L010", "L020", "L030", "L040",
                       "L050", "L060", "L070", "L080", "L090", "L100")
La_Selva <- "La Selva Biological Station"
Barro <- "Barro Colorado Island"
Tiputini <- "Tiputini Biodiversity Station"
La_Selva_country <- "Costa Rica"
Barro_country <- "Panama"
Tiputini_country <- "Ecuador"
title_prefix <- "soil sample from "
La_Selva_title <- str_c(title_prefix, La_Selva, ", ", La_Selva_country)
Barro_title <- str_c(title_prefix, Barro, ", ", Barro_country)
Tiputini_title <- str_c(title_prefix, Tiputini, ", ", Tiputini_country)
output <- "neotrop_samples.tsv"
```

```{r}
load_list_of_read_files <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(sample = str_remove(string = Submitted_files,
                                    pattern = "_lib.*"),
               sample = str_remove(string = sample,
                                    pattern = "NG-[0-9]+_")) %>%
        filter(str_detect(string = Submitted_files,
                          pattern = "_1_2\\.",
                          negate = TRUE))
}


load_per_sample_coordinates <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        mutate(northing = as.character(northing),
               sample_description = str_glue(
                   "Sampling position of {sample}",
                   "in UTM coordinates is zone: {zone},",
                   "northing: {northing}, easting: {easting}.",
                   .sep = " ")
               )
}


load_raw_metadata <- function(filename){
    here::here("data", filename) %>%
        read_tsv(show_col_types = FALSE) %>%
        filter(project == project_name_in_eukbank) %>%
        select(-c(project, notus, nreads, size_fraction_lower_threshold,
                  size_fraction_upper_threshold))
}


. %>%
    mutate(sample = case_when(
               sample == "T185_186" ~ "T185_T186",
               sample == "L137_L13" ~ "L137_L138",
               TRUE                 ~ sample),
           collection_date = case_when(
               sample == "T185_T186" ~ "2013-10-01",
               sample == "T199_T200" ~ "2013-10-01",
               sample == "L137_L138" ~ "2013-06-01",
               TRUE                  ~ collection_date),
           longitude = case_when(
               sample == "T185_T186" ~ -76.1690818455638,
               sample == "T199_T200" ~ -76.1583931324974,
               sample == "L137_L138" ~ -84.0081054902495,
               TRUE                  ~ longitude),
           latitude = case_when(
               sample == "T185_T186" ~ -0.622442092820539,
               sample == "T199_T200" ~ -0.61982581193797,
               sample == "L137_L138" ~ 10.4242830027919,
               TRUE                  ~ latitude)
           ) -> fix_missing_data
```


### get a list of available read files (sff and fastq)

```{r}
load_list_of_read_files(list_of_read_files) -> read_file_names
```


Some samples were sequenced twice (Roche 454 and Illumina
MiSeq). After extracting sample names from read file names, we expect
to find 154 unique sample names.

```{r}
read_file_names %>%
    distinct(sample) %>%
    pull() %>%
    length() -> n_unique_samples
```

There are `r prettyNum(n_unique_samples, scientific=FALSE,
big.mark=",")` unique sample names.


### get per-sample coordinates

Using the Universal Transverse Mercator (UTM) coordinate system:

```{r}
load_per_sample_coordinates(coordinates_file) -> sample_cordinates
```


# load metadata

```{r}
load_raw_metadata(eukbank_metadata_file) %>%
    fix_missing_data -> metadata
```


## check

Are sample lists equal?

```{r}
setequal(x = metadata %>%
             distinct(sample),
         y = read_file_names %>%
             distinct(sample))
```

after fixing two sample names, the metadata table and the list of
available read files now differ.


## create the 'sample description' field

generate sample description for double samples:

```{r}
. %>%
    select(sample) %>%
    filter(str_detect(string = sample, pattern = "_")) %>%
    mutate(sample_pair = sample) %>%
    separate_rows(sample, sep = "_") %>%
    left_join(sample_cordinates %>%
              select(sample, sample_description),
              by = "sample") %>%
    select(-sample) %>%
    group_by(sample_pair) %>%
    mutate(position = row_number(sample_pair)) %>%
    ungroup() %>%
    pivot_wider(names_from = position, values_from = sample_description) %>%
    unite(2, 3, col = "sample_description", sep = " ") %>%
    mutate(sample_description = as.character(sample_description)) %>%
    rename(sample = sample_pair) -> sample_descriptions_for_double_samples
```

bind with sample descriptions for single samples:

```{r}
bind_rows(metadata %>%
          sample_descriptions_for_double_samples,
          sample_cordinates %>%
          select(sample, sample_description)
          ) -> sample_descriptions
```


## ENA UniEuk_EukBank Checklist

Create and order columns as specified in the [checklist
ERC000040](https://www.ebi.ac.uk/ena/browser/view/ERC000040):

```{r}
. %>%
    mutate(tax_id = 342097,
           scientific_name = "uncultured soil eukaryote",
           sample_alias = sample,
           sample_title = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva_title,
               str_starts(string = sample, pattern = "B") ~ Barro_title,
               str_starts(string = sample, pattern = "T") ~ Tiputini_title,
               TRUE ~ "unknown"
           ),
           sample_description = as.character(sample_description),
           `target gene` = "18S SSU rRNA",
           `target subfragment` = "V4",
           `pcr primers` = primers,
           isolation_source	= "forest top-soil",
           collected_by = "Micah Dunthorn",
           collection_date = lubridate::parse_date_time(collection_date, "ymd"),
           collection_date = remove_days(collection_date),
           `collection date` = collection_date,
           `geographic location (country and/or sea)` = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva_country,
               str_starts(string = sample, pattern = "B") ~ Barro_country,
               str_starts(string = sample, pattern = "T") ~ Tiputini_country,
               TRUE ~ "unknown"
           ),
           `geographic location (latitude)` = latitude,
           `geographic location (longitude)` = longitude,
           `geographic location (region and locality)` = case_when(
               str_starts(string = sample, pattern = "L") ~ La_Selva,
               str_starts(string = sample, pattern = "B") ~ Barro,
               str_starts(string = sample, pattern = "T") ~ Tiputini,
               TRUE ~ "unknown"
           ),
           depth = 0,
           `broad-scale environmental context` = biome,
           `local environmental context` = feature,
           `environmental medium` = material,
           `sample collection device or method` = "spoon",
           environmental_sample = "Yes",
           `Further Details` = "http://unieuk.org/project/",
           .before = 1) -> create_eukbank_columns
```

Which level of precision should we choose?

- `tax_id` = 100272, `scientific_name` = "uncultured eukaryote"
- `tax id` = 342097, `scientific_name` = "uncultured soil eukaryote"


### Illumina and Roche 454 samples (universal primers)

The primers are:

- V4F = CCAGCASCYGCGGTAATTCC
- V4R = ACTTTCGTTCTTGATYRA

Also known as:

- TAReuk454FWD1 (Forward): 5‘-CCAGCA(G/C)C(C/T)GCGGTAATTCC-3‘
- TAReukRev3 (Reverse): 5‘-ACTTTCGTTCTTGAT(C/T)(A/-G)A-3‘

Standard master mix (in μl):

- 19,75 H20
- 2,5 10x Buffer
- 0,5 dNTP
- 0,5 V4F
- 0,5 V4R
- 0,25 polymerase
- 1 template

Cycling conditions:

(with Hot Start Taq from Qiagen)

1) 95°C 5 min
2) 94°C 30 sec
3) 47°C 45 sec
4) 72°C 60 sec; 2-4 = 29x
5) 72°C 5 min


```{r}
metadata %>%
    left_join(x = ., y = sample_descriptions, by = "sample") %>%
    create_eukbank_columns %>%
    relocate(sample_description, .after = sample_title) %>%
    relocate(depth, .after = `geographic location (region and locality)`) %>%
    select(tax_id:`Further Details`) %>%
    arrange(sample_alias) -> universal_primer_samples
```


### Roche 454 samples (ciliate-specific primers, followed by universal primers)

20 samples, using a two step amplification:

1) ciliate specific primers,
2) universal V4 primers

The primers are:

- Cil_F: 5’ - TGG TAG TGT ATT GGA CWA CCA -3’

Cil R (an equimolar mix of three different primers):

- 5’ - TCT GAT CGT CTT TGA TCC CTT A – 3’
- 5’ - TCT RAT CGT CTT TGA TCC CCT A – 3’
- 5’ - TCT GAT TGT CTT TGA TCC CCT A – 3’

Cycling conditions:

(with Hot Start Taq from Qiagen)

1) 95°C 5 min
2) 94°C 30 sec
3) 56°C 30 sec
4) 72°C 60 sec; 2-4 = 25x
5) 72°C 10 min

followed by an amplification with universal 18S SSU rRNA V4 primers
(see above).

```{r}
universal_primer_samples %>%
    filter(sample_alias %in% roche_454_samples) %>%
    mutate(tax_id = 187299,
           scientific_name = "uncultured ciliate",
           `Further Details` = "") -> ciliate_specific_samples
```

Note that these samples are marked as *uncultured ciliate*, but might
contain sequences from other taxonomic groups.


## save results

```{r}
bind_rows(universal_primer_samples, ciliate_specific_samples) %>%
    write_tsv(here::here("results", output))
```


***

```{r}
sessionInfo()
rm(list = ls())
```
